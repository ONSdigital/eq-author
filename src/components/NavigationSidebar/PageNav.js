import React from "react";
import styled from "styled-components";

import CustomPropTypes from "custom-prop-types";
import { CSSTransition, TransitionGroup } from "react-transition-group";

import PageNavItem from "./PageNavItem";

const duration = 300;

const NavList = styled.ol`
  padding: 0;
  margin: 0;
  list-style: none;
`;

const PageNav = ({ section, questionnaire }) => (
  <TransitionGroup component={NavList}>
    {section.pages.map(page => (
      <CSSTransition key={page.id} timeout={duration} classNames="page">
        <PageNavItem
          title={page.title}
          pageId={page.id}
          sectionId={section.id}
          questionnaireId={questionnaire.id}
        />
      </CSSTransition>
    ))}
  </TransitionGroup>
);

PageNav.propTypes = {
  questionnaire: CustomPropTypes.questionnaire.isRequired,
  section: CustomPropTypes.section.isRequired
};

export default PageNav;
