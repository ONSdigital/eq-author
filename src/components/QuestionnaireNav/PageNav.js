import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { getLink } from "utils/UrlUtils";
import Tooltip from "components/Tooltip";
import CustomPropTypes from "custom-prop-types";
import { TransitionGroup, CSSTransition } from "react-transition-group";
import { NavLink } from "react-router-dom";
import { colors } from "constants/theme";
import getTextFromHTML from "utils/getTextFromHTML";
import HoverDeleteButton from "./HoverDeleteButton";

const duration = 300;

export const StyledPageItem = styled.li`
  padding: 0;
  margin: 0;
  font-weight: 500;
  display: flex;
  align-items: center;
  z-index: ${props => props.index};
  position: relative;

  &.page-enter,
  &.page-exit.page-exit-active {
    opacity: 0;
    height: 0;
    transform: translateX(-20px);
  }

  &.page-exit,
  &.page-enter.page-enter-active {
    opacity: 1;
    height: 2em;
    transform: translateX(0);
  }

  &.page-enter-active {
    transition: height ${duration / 2}ms ease-out,
      opacity ${duration / 2}ms ease-out ${duration / 2}ms,
      transform ${duration / 2}ms ease-out ${duration / 2}ms;
  }

  &.page-exit-active {
    transition: opacity ${duration / 2}ms ease-out,
      transform ${duration / 2}ms ease-out,
      height ${duration / 2}ms ease-out ${duration / 2}ms;
  }
`;

const Link = styled(NavLink)`
  text-decoration: none;
  color: ${colors.text};
  font-size: 0.75em;
  padding: 0.7em 2.5em 0.7em 0.9em;
  display: block;
  flex: 1 1 auto;
  height: 100%;
  position: relative;
  overflow: hidden;
  transition: opacity 100ms ease-out;

  &::before {
    opacity: 0;
    content: "";
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 0;
    transition: all 200ms ease-out;
    background: ${colors.borders};
    z-index: 1;
  }

  /* stylelint-disable */
  ${StyledPageItem}:hover & {
    &::before {
      opacity: 0.5;
      width: 100%;
    }
  }
  /* stylelint-enable */

  &.selected {
    &::before {
      opacity: 1 !important;
      width: 100%;
    }
  }

  &[aria-disabled="true"] {
    pointer-events: none;
    opacity: 0.5;
  }
`;

export const LinkText = styled.span`
  display: inline-block;
  vertical-align: middle;
  width: 100%;
  position: relative;
  z-index: 2;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  opacity: ${({ fade }) => (fade ? 0.5 : 1)};
`;

const NavList = styled.ol`
  padding: 0;
  margin: 0;
  list-style: none;
`;

export const DeleteButton = styled(HoverDeleteButton)`
  top: 0.1em;
  /* stylelint-disable */
  ${StyledPageItem}:hover &,
  ${Link}:focus + &,
  &:focus {
    opacity: 1;
    transform: translateX(0);
  }
  /* stylelint-enable */
`;

export class PageNavItem extends React.Component {
  state = {
    isDeleting: false
  };

  static propTypes = {
    sectionId: PropTypes.string.isRequired,
    questionnaireId: PropTypes.string.isRequired,
    pageId: PropTypes.string.isRequired,
    pageNumber: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    onDelete: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired
  };

  handleDelete = () => {
    const { sectionId, pageId, onDelete } = this.props;

    this.setState({ isDeleting: true });

    onDelete(sectionId, pageId).catch(() => {
      this.setState({ isDeleting: false });
    });
  };

  render() {
    const {
      sectionId,
      questionnaireId,
      pageId,
      title,
      pageNumber,
      index
    } = this.props;

    return (
      <StyledPageItem index={index}>
        <Link
          to={getLink(questionnaireId, sectionId, pageId)}
          aria-disabled={parseInt(pageId, 10) < 0}
          activeClassName="selected"
        >
          <LinkText fade={this.state.isDeleting}>
            {pageNumber} {getTextFromHTML(title) || "Page Title"}
          </LinkText>
        </Link>
        <Tooltip content="Delete page">
          <DeleteButton
            type="button"
            aria-label="Delete page"
            onClick={this.handleDelete}
          >
            Ã—
          </DeleteButton>
        </Tooltip>
      </StyledPageItem>
    );
  }
}

const PageNav = ({ section, questionnaire, onDelete }) => (
  <TransitionGroup component={NavList}>
    {section.pages.map((page, i, pages) => {
      const pageNumber = `${section.number}.${i + 1}`;
      return (
        <CSSTransition key={page.id} timeout={duration} classNames="page">
          <PageNavItem
            title={page.title}
            pageNumber={pageNumber}
            pageId={page.id}
            sectionId={section.id}
            questionnaireId={questionnaire.id}
            onDelete={onDelete}
            index={pages.length - i}
          />
        </CSSTransition>
      );
    })}
  </TransitionGroup>
);

PageNav.propTypes = {
  questionnaire: CustomPropTypes.questionnaire.isRequired,
  section: CustomPropTypes.section.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default PageNav;
