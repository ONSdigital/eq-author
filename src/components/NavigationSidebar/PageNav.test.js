import React from "react";
import { shallow } from "enzyme";
import PageNav, {
  DeleteButton,
  PageNavItem,
  LinkText,
  StyledPageItem
} from "./PageNav";

describe("PageNav", () => {
  let component, handleDelete;

  const questionnaire = { id: "1", title: "Questionnaire" };
  const page = { id: "2", title: "Page" };
  const section = { id: "3", title: "Section", pages: [page], number: 1 };

  beforeEach(() => {
    handleDelete = jest.fn(() => Promise.resolve());

    component = shallow(
      <PageNav
        questionnaire={questionnaire}
        section={section}
        onDelete={handleDelete}
      />
    );
  });

  it("should render", () => {
    expect(component).toMatchSnapshot();
  });

  describe("PageNavItem", () => {
    let pageNavComponent;

    beforeEach(() => {
      pageNavComponent = component.find(PageNavItem).dive();
    });

    it("should render", () => {
      expect(pageNavComponent).toMatchSnapshot();
    });

    describe("when delete button clicked", () => {
      it("should invoke callback when delete button clicked", () => {
        pageNavComponent.find(DeleteButton).simulate("click");
        expect(handleDelete).toHaveBeenCalledWith("3", "2");
      });

      it("should set isDeleting state to true", () => {
        pageNavComponent.find(DeleteButton).simulate("click");
        expect(pageNavComponent.state().isDeleting).toBe(true);
      });

      it("should set isDeleting state false in event of failure", () => {
        const handleCatch = jest.fn(arg => arg());
        const handleDelete = () => ({ catch: handleCatch });

        pageNavComponent.setProps({ onDelete: handleDelete });
        pageNavComponent.find(DeleteButton).simulate("click");

        expect(handleCatch).toHaveBeenCalled();
        expect(pageNavComponent.state().isDeleting).toBe(false);
      });
    });
  });

  describe("LinkText", () => {
    it("should be opaque", () => {
      const wrapper = shallow(<LinkText />);
      expect(wrapper).toHaveStyleRule("opacity", "1");
    });

    it("should be partially transparent if fade prop passed", () => {
      const wrapper = shallow(<LinkText fade />);
      expect(wrapper).toHaveStyleRule("opacity", "0.5");
    });
  });

  describe("StyledPageItem", () => {
    it("should have configurable z-index", () => {
      const wrapper = shallow(<StyledPageItem index={123} />);
      expect(wrapper).toHaveStyleRule("z-index", "123");
    });
  });

  describe("Fix keyboard focus scrolling", () => {
    let querySelector;
    let wrapper;
    let props;
    let mockScrollPane;

    beforeEach(() => {
      querySelector = jest.spyOn(document, "querySelector");

      props = {
        sectionId: section.id,
        questionnaireId: questionnaire.id,
        pageId: page.id,
        pageNumber: page.id,
        title: "Title",
        onDelete: handleDelete,
        index: 0
      };

      mockScrollPane = document.createElement("div");

      wrapper = shallow(<PageNavItem {...props} />);
    });

    afterEach(() => {
      jest.restoreAllMocks();
    });

    it("should set the scrollLeft to 0", () => {
      querySelector.mockReturnValue(mockScrollPane);
      wrapper.find(DeleteButton).simulate("focus");
      expect(querySelector).toHaveBeenCalledWith(
        '[class*="NavigationScrollPane"]'
      );
      expect(mockScrollPane.scrollLeft).toEqual(0);
    });
  });
});
