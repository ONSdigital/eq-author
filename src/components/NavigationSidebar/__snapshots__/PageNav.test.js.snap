// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageNav should render 1`] = `
<TransitionGroup
  childFactory={[Function]}
  component={[Function]}
>
  <CSSTransition
    classNames="page"
    key="2"
    timeout={300}
  >
    <PageNavItem
      pageId="2"
      questionnaireId="1"
      sectionId="3"
      title="Page"
    />
  </CSSTransition>
</TransitionGroup>
`;
